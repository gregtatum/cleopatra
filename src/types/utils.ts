import { $Call } from "utility-types";

/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */





/*
 * A not-void type
 * https://flow.org/try/#0C4TwDgpgBAcg9sAanAlgEygXigKClAHygDsBXAWwCMIAnPQqAZ2BpWIHN6jK44AbCAENiXEqT59RAQRo1BIADzCQAPlEB5SgCsIAY2ABuHDgBmpYvpRxiUMxYUAVAFywEydCoAUAN0F9SEC4OAJRBUADe9DQQwKQ0Nr7+EEYAvsZ2up4AjMFGGZ4ARMysHAW5puaZ4ba8LgDklII0dVAp5flkEu2VngDaWQA0UABMALrdFp7maBAmbBBoE5nlQA
 */
export type NotVoidOrNull = number | string | boolean | Array<any> | Object;

export type ExtractReturnType = <V>(arg0: (...args: any[]) => V) => V;

/**
 * This is a utility type that extracts the return type of a function.
 */
export type $ReturnType<Fn> = $Call<ExtractReturnType, Fn>;