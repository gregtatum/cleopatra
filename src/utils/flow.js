/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import type { TabSlug } from '../types/actions';
import type { TransformType } from '../types/transforms';

/**
 * This file contains utils that help Flow understand things better. Occasionally
 * statements can be logically equivalent, but Flow infers them in a specific way. Most
 * of the time tweaks can be done by editing the type system, but occasionally functions
 * are needed to get the desired result.
 */

/**
 * This function can be run as the default arm of a switch statement to ensure exhaustive
 * checking of a given type. It relies on an assumption that all cases will be handled
 * and the input to the function will be empty. This function hopefully makes that check
 * more readable.
 */
export function assertExhaustiveCheck(notValid: empty): void {
  throw new Error(`There was an unhandled case for the value: "${notValid}"`);
}

/**
 * Immutably update an object through Object.assign, but retain the original
 * type information of the object. Flow will occasionally throw errors when
 * inferring what is going on with Object.assign.
 */
export function immutableUpdate<T: Object>(object: T, ...rest: Object[]): T {
  return Object.assign({}, object, ...rest);
}

/**
 * This function takes a string and returns either a valid TabSlug or null, this doesn't
 * throw an error so that any arbitrary string can be converted, e.g. from a URL.
 */
export function toValidTabSlug(tabSlug: any): TabSlug | null {
  const coercedTabSlug = (tabSlug: TabSlug);
  switch (coercedTabSlug) {
    case 'calltree':
    case 'stack-chart':
    case 'marker-chart':
    case 'marker-table':
    case 'flame-graph':
      return coercedTabSlug;
    default: {
      // The coerced type SHOULD be empty here. If in reality we get
      // here, then it's not a valid transform type, so return null.
      (coercedTabSlug: empty); // eslint-disable-line no-unused-expressions
      return null;
    }
  }
}

/**
 * This function will take an arbitrary string, and try to convert it to a valid
 * TransformType.
 */
export function convertToTransformType(type: string): TransformType | null {
  // Coerce this into a TransformType even if it's not one.
  const coercedType = ((type: any): TransformType);
  switch (coercedType) {
    // Exhaustively check each TransformType. The default arm will assert that
    // we have been exhaustive.
    case 'merge-call-node':
    case 'merge-function':
    case 'focus-subtree':
    case 'focus-function':
    case 'collapse-resource':
    case 'collapse-direct-recursion':
    case 'collapse-function-subtree':
    case 'drop-function':
      return coercedType;
    default: {
      // The coerced type SHOULD be empty here. If in reality we get
      // here, then it's not a valid transform type, so return null.
      (coercedType: empty); // eslint-disable-line no-unused-expressions
      return null;
    }
  }
}

/**
 * This function will take an arbitrary string, and will turn it into a TransformType
 * it will throw an error if an invalid type was passed to it.
 */
export function ensureIsTransformType(type: string): TransformType {
  const assertedType = convertToTransformType(type);
  if (!assertedType) {
    throw new Error(
      `Attempted to assert that "${type}" is a valid TransformType, and it was not.`
    );
  }
  return assertedType;
}
