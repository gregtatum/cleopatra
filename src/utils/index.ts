/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */




/**
 * Firefox has issues switching quickly between fill style colors, as the CSS color
 * is fully parsed each time it is set. As a mitigation, provide a class that only
 * switches the color when it's really needed.
 */
export class FastFillStyle {

  _ctx: CanvasRenderingContext2D;
  _previousFillColor: string;

  constructor(ctx: CanvasRenderingContext2D) {
    this._ctx = ctx;
    this._previousFillColor = '';
  }

  set(fillStyle: string) {
    if (fillStyle !== this._previousFillColor) {
      this._ctx.fillStyle = fillStyle;
      this._previousFillColor = fillStyle;
    }
  }
}

/**
 * Perform a simple shallow object equality check.
 */
export function objectShallowEquals(a: Object, b: Object): boolean {
  let aLength = 0;
  let bLength = 0;
  for (const key in a) {
    if (Object.prototype.hasOwnProperty.call(a, key)) {
      aLength++;
      if (a[key] !== b[key]) {
        return false;
      }
    }
  }
  for (const key in b) {
    if (Object.prototype.hasOwnProperty.call(b, key)) {
      bLength++;
    }
  }
  return aLength === bLength;
}

/**
 * Don't completely trust URLs that get displayed to the screen. These could
 * be extraordinarily long data-uris. This function could also be updated to
 * return a nicely formatted React span.
 */
export function displayNiceUrl(rawUrl: string): string {
  if (rawUrl.length < 200) {
    return rawUrl;
  }
  return `${rawUrl.slice(0, 150)}...${rawUrl.slice(-20)}`;
}