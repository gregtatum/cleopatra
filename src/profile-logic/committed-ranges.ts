/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */


import { StartEndRange } from "../types/units";

/**
 * Users can make preview range selections on the profile, and then can commit these
 * to drill down into a profile. This file contains functions for working with these
 * committed ranges.
 */

/**
 * Parse URL encoded committed ranges with the form: "start-end~start-end", where
 * `start` and `end` are positive or negative float numbers.
 */
export function parseCommittedRanges(stringValue: string = ''): StartEndRange[] {
  if (!stringValue) {
    return [];
  }
  return stringValue.split('~').map(s => {
    const m = s.match(/(-?[0-9.]+)_(-?[0-9.]+)/);
    if (!m) {
      return { start: 0, end: 1000 };
    }

    const m1 = Number(m[1]);
    let m2 = Number(m[2]);

    if (m2 === m1) {
      // Ensure that the duration of the range is non-zero.
      m2 = m2 + 0.0001;
    }
    return { start: m1 * 1000, end: m2 * 1000 };
  });
}

/**
 * Stringify committed ranges into the following form: "start-end~start-end", where
 * `start` and `end` are float numbers.
 */
export function stringifyCommittedRanges(arrayValue: StartEndRange[] = []): string {
  return arrayValue.map(({
    start,
    end
  }) => {
    const startStr = (start / 1000).toFixed(4);
    const endStr = (end / 1000).toFixed(4);
    return `${startStr}_${endStr}`;
  }).join('~');
}

export function getFormattedTimeLength(length: number): string {
  if (length >= 10000) {
    return `${(length / 1000).toFixed(0)} sec`;
  }
  if (length >= 1000) {
    return `${(length / 1000).toFixed(1)} sec`;
  }
  return `${length.toFixed(0)} ms`;
}

export function getCommittedRangeLabels(committedRanges: StartEndRange[]): string[] {
  const labels = committedRanges.map(range => getFormattedTimeLength(range.end - range.start));
  return labels;
}