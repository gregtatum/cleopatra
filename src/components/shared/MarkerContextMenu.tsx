/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import React, { PureComponent } from 'react';
import { MenuItem } from 'react-contextmenu';
import ContextMenu from '../shared/ContextMenu';
import explicitConnect from '../../utils/connect';
import {
  setContextMenuVisibility,
  updatePreviewSelection,
} from '../../actions/profile-view';
import {
  getPreviewSelection,
  getCommittedRange,
} from '../../selectors/profile';
import { getRightClickedMarkerInfo } from '../../selectors/right-clicked-marker';
import copy from 'copy-to-clipboard';

import type { Marker } from '../../types/profile-derived';
import type { StartEndRange } from '../../types/units';
import type {
  PreviewSelection,
  ImplementationFilter,
} from '../../types/actions';
import type { ConnectedProps } from '../../utils/connect';
import { getImplementationFilter } from '../../selectors/url-state';
import type { IndexIntoStackTable, Thread } from '../../types/profile';
import { filterCallNodePathByImplementation } from '../../profile-logic/transforms';
import {
  convertStackToCallNodePath,
  getFuncNamesAndOriginsForPath,
} from '../../profile-logic/profile-data';
import { getMarkerFullDescription } from '../../profile-logic/marker-data';
import { getThreadSelectors } from '../../selectors/per-thread';

type StateProps = {|
  +previewSelection: PreviewSelection,
  +committedRange: StartEndRange,
  +rightClickedMarker: Marker | null,
  +thread: Thread | null,
  +implementationFilter: ImplementationFilter,
|};

type DispatchProps = {|
  +updatePreviewSelection: typeof updatePreviewSelection,
  +setContextMenuVisibility: typeof setContextMenuVisibility,
|};

type Props = ConnectedProps<{||}, StateProps, DispatchProps>;

class MarkerContextMenu extends PureComponent<Props> {
  setStartRange = () => {
    const {
      rightClickedMarker,
      updatePreviewSelection,
      previewSelection,
      committedRange,
    } = this.props;

    if (rightClickedMarker === null) {
      return;
    }

    const selectionEnd = previewSelection.hasSelection
      ? previewSelection.selectionEnd
      : committedRange.end;

    updatePreviewSelection({
      hasSelection: true,
      isModifying: false,
      selectionStart: rightClickedMarker.start,
      selectionEnd,
    });
  };

  setEndRange = () => {
    const {
      rightClickedMarker,
      updatePreviewSelection,
      committedRange,
      previewSelection,
    } = this.props;

    if (rightClickedMarker === null) {
      return;
    }

    const selectionStart = previewSelection.hasSelection
      ? previewSelection.selectionStart
      : committedRange.start;

    updatePreviewSelection({
      hasSelection: true,
      isModifying: false,
      selectionStart,
      // For markers without a duration, add an arbitrarily small bit of time at
      // the end to make sure the selected marker doesn't disappear from view.
      selectionEnd:
        rightClickedMarker.start + (rightClickedMarker.dur || 0.0001),
    });
  };

  setRangeByDuration = () => {
    const { rightClickedMarker, updatePreviewSelection } = this.props;

    if (rightClickedMarker === null) {
      return;
    }

    if (this._isZeroDurationMarker(rightClickedMarker)) {
      return;
    }

    updatePreviewSelection({
      hasSelection: true,
      isModifying: false,
      selectionStart: rightClickedMarker.start,
      selectionEnd: rightClickedMarker.start + rightClickedMarker.dur,
    });
  };

  _isZeroDurationMarker(marker: ?Marker): boolean {
    return !marker || !marker.dur;
  }

  _convertStackToString(stack: IndexIntoStackTable): string {
    const { thread, implementationFilter } = this.props;

    if (thread === null) {
      return '';
    }

    const callNodePath = filterCallNodePathByImplementation(
      thread,
      implementationFilter,
      convertStackToCallNodePath(thread, stack)
    );

    const funcNamesAndOrigins = getFuncNamesAndOriginsForPath(
      callNodePath,
      thread
    );
    return funcNamesAndOrigins
      .map(({ funcName, origin }) => `${funcName} [${origin}]`)
      .join('\n');
  }

  copyMarkerJSON = () => {
    const { rightClickedMarker } = this.props;

    if (rightClickedMarker === null) {
      return;
    }

    copy(JSON.stringify(rightClickedMarker, null, 2));
  };

  copyMarkerDescription = () => {
    const { rightClickedMarker } = this.props;

    if (rightClickedMarker === null) {
      return;
    }

    copy(getMarkerFullDescription(rightClickedMarker));
  };

  copyMarkerCause = () => {
    const { rightClickedMarker } = this.props;

    if (rightClickedMarker === null) {
      return;
    }

    if (rightClickedMarker.data && rightClickedMarker.data.cause) {
      const stack = this._convertStackToString(
        rightClickedMarker.data.cause.stack
      );
      if (stack) {
        copy(stack);
      } else {
        copy(
          'The stack is empty because all of its frames are filtered out by the implementation filter. Switch the implementation filter in the call tree to see more frames.'
        );
      }
    }
  };

  copyUrl = () => {
    const { rightClickedMarker } = this.props;

    if (rightClickedMarker === null) {
      return;
    }

    if (rightClickedMarker.data && rightClickedMarker.data.type === 'Network') {
      copy(rightClickedMarker.data.URI);
    }
  };

  // Using setTimeout here is a bit complex, but is necessary to make the menu
  // work fine when we want to display it somewhere when it's already open
  // somewhere else.
  // This is the order of events in such a situation:
  // 0. The menu is open somewhere, it means the user right clicked somewhere
  //     previously, and as a result some marker has the "right clicked" status.
  // 1. The user right clicks on another marker. This is actually happening in
  //    several events, the first event is "mousedown": this is where our own
  //    components react for right click (both our TreeView and our charts)
  //    and thus this is when the "right clicked" item is set in our store. BTW
  //    this triggers a rerender of this component.
  // 2. Then the event "mouseup" happens but we don't do anything for it for right
  //    clicks.
  // 3. Then the event "contextmenu" is triggered. This is the event that the
  //    context menu library reacts to: first it closes the previous menu, then
  //    opens the new one. This means that `_onHide` is called first for the
  //    first menu, then `_onShow` for the second menu.
  //    The problem here is that the call to `setContextMenuVisibility` we do in
  //    `onHide` resets the value for the "right clicked" item. This is normally
  //    what we want when the user closes the menu, but in this case where the
  //    menu is still open but for another node, we don't want to reset this
  //    value which was set earlier when handling the "mousedown" event.
  //    To avoid this problem we use this `setTimeout` call to delay the reset
  //    just a bit, just in case we get a `_onShow` call right after that.
  _hidingTimeout: TimeoutID | null = null;

  _onHide = () => {
    this._hidingTimeout = setTimeout(() => {
      this._hidingTimeout = null;
      this.props.setContextMenuVisibility(false);
    });
  };

  _onShow = () => {
    clearTimeout(this._hidingTimeout);
    this.props.setContextMenuVisibility(true);
  };

  render() {
    const { rightClickedMarker } = this.props;

    if (rightClickedMarker === null) {
      return null;
    }

    return (
      <ContextMenu
        id="MarkerContextMenu"
        onShow={this._onShow}
        onHide={this._onHide}
      >
        <MenuItem onClick={this.setStartRange}>
          Set selection start time here
        </MenuItem>
        <MenuItem onClick={this.setEndRange}>
          Set selection end time here
        </MenuItem>
        <MenuItem
          onClick={this.setRangeByDuration}
          disabled={this._isZeroDurationMarker(rightClickedMarker)}
        >
          Set selection from duration
        </MenuItem>
        <MenuItem onClick={this.copyMarkerDescription}>Copy</MenuItem>
        {rightClickedMarker.data && rightClickedMarker.data.cause ? (
          <MenuItem onClick={this.copyMarkerCause}>Copy marker cause</MenuItem>
        ) : null}
        {rightClickedMarker.data &&
        rightClickedMarker.data.type === 'Network' ? (
          <MenuItem onClick={this.copyUrl}>Copy URL</MenuItem>
        ) : null}
        <MenuItem onClick={this.copyMarkerJSON}>Copy marker JSON</MenuItem>
      </ContextMenu>
    );
  }
}

export default explicitConnect<{||}, StateProps, DispatchProps>({
  mapStateToProps: state => {
    const rightClickedMarkerInfo = getRightClickedMarkerInfo(state);

    let rightClickedMarker = null;
    let thread = null;

    if (rightClickedMarkerInfo !== null) {
      const selectors = getThreadSelectors(rightClickedMarkerInfo.threadIndex);

      rightClickedMarker = selectors.getRightClickedMarker(state);
      thread = selectors.getThread(state);
    }

    return {
      previewSelection: getPreviewSelection(state),
      committedRange: getCommittedRange(state),
      implementationFilter: getImplementationFilter(state),
      rightClickedMarker,
      thread,
    };
  },
  mapDispatchToProps: { updatePreviewSelection, setContextMenuVisibility },
  component: MarkerContextMenu,
});
