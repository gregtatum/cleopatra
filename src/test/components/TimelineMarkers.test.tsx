/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';

// This module is mocked.
import copy from 'copy-to-clipboard';

import {
  TimelineMarkersOverview,
  MIN_MARKER_WIDTH,
} from '../../components/timeline/Markers';
import MarkerContextMenu from '../../components/shared/MarkerContextMenu';
import { overlayFills } from '../../profile-logic/marker-styles';
import { render, fireEvent } from 'react-testing-library';
import { Provider } from 'react-redux';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileWithMarkers } from '../fixtures/profiles/processed-profile';
import {
  getBoundingBox,
  getMouseEvent,
  addRootOverlayElement,
  removeRootOverlayElement,
} from '../fixtures/utils';
import mockRaf from '../fixtures/mocks/request-animation-frame';
import { ensureExists } from '../../utils/flow';

import type { CssPixels } from '../../types/units';

function setupWithMarkers({ rangeStart, rangeEnd }, ...markersPerThread) {
  const flushRafCalls = mockRaf();
  const ctx = mockCanvasContext();

  jest
    .spyOn(HTMLCanvasElement.prototype, 'getContext')
    .mockImplementation(() => ctx);

  // Ideally we'd want this only on the Canvas and on ChartViewport, but this is
  // a lot easier to mock this everywhere.
  jest
    .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
    .mockImplementation(() => getBoundingBox(200, 300));

  function flushRafTwice() {
    // We need to flush twice since when the first flush is run, it
    // will request more code to be run in later animation frames.
    flushRafCalls();
    flushRafCalls();
  }

  function flushDrawLog() {
    return ctx.__flushDrawLog();
  }

  const profile = getProfileWithMarkers(...markersPerThread);

  const renderResult = render(
    <Provider store={storeWithProfile(profile)}>
      <>
        <TimelineMarkersOverview
          rangeStart={rangeStart}
          rangeEnd={rangeEnd}
          threadIndex={0}
          onSelect={() => {}}
        />
        <MarkerContextMenu />
      </>
    </Provider>
  );

  flushRafTwice();

  function getContextMenu() {
    return ensureExists(
      renderResult.container.querySelector('.react-contextmenu'),
      `Couldn't find the context menu.`
    );
  }

  function clickOnMenuItem(stringOrRegexp) {
    fireEvent.click(renderResult.getByText(stringOrRegexp));
  }

  function fireMouseEvent(eventName, options) {
    fireEvent(
      ensureExists(
        renderResult.container.querySelector('canvas'),
        `Couldn't find the canvas element`
      ),
      getMouseEvent(eventName, options)
    );
  }

  function getPositioningOptions({ x, y }) {
    // These positioning options will be sent to all our mouse events. Note
    // that the values aren't really consistent, especially offsetY and
    // pageY shouldn't be the same, but in the context of our test this will
    // be good enough.
    // pageX/Y values control the position of the tooltip so it's not super
    // important.
    // offsetX/Y are more important as they're used to find which node is
    // actually clicked.
    const positioningOptions = {
      offsetX: x,
      offsetY: y,
      clientX: x,
      clientY: y,
      pageX: x,
      pageY: y,
    };

    return positioningOptions;
  }

  // Note to a future developer: the x/y values can be derived from the
  // array returned by ctx.__flushDrawLog().
  function rightClick(where: { x: CssPixels, y: CssPixels }) {
    const positioningOptions = getPositioningOptions(where);
    const clickOptions = {
      ...positioningOptions,
      button: 2,
      buttons: 2,
    };

    // Because different components listen to different events, we trigger
    // all the right events, to be as close as possible to the real stuff.
    fireMouseEvent('mousemove', positioningOptions);
    fireMouseEvent('mousedown', clickOptions);
    fireMouseEvent('mouseup', clickOptions);
    fireMouseEvent('contextmenu', clickOptions);

    flushRafTwice();
  }

  function mouseOver(where: { x: CssPixels, y: CssPixels }) {
    const positioningOptions = getPositioningOptions(where);
    fireMouseEvent('mousemove', positioningOptions);
    flushRafTwice();
  }

  return {
    flushDrawLog,
    flushRafTwice,
    rightClick,
    mouseOver,
    getContextMenu,
    clickOnMenuItem,
    ...renderResult,
  };
}

describe('TimelineMarkers', function() {
  it('renders correctly', () => {
    window.devicePixelRatio = 1;

    const { container, flushDrawLog } = setupWithMarkers(
      { rangeStart: 0, rangeEnd: 15 },
      [
        ['Marker A', 0, { startTime: 0, endTime: 10 }],
        ['Marker B', 0, { startTime: 0, endTime: 10 }],
        ['Marker C', 5, { startTime: 5, endTime: 15 }],
        [
          'BHR-detected hang',
          5,
          { type: 'BHR-detected hang', startTime: 2, endTime: 13 },
        ],
        [
          'LongTask',
          6,
          {
            type: 'MainThreadLongTask',
            category: 'LongTask',
            startTime: 2,
            endTime: 6,
          },
        ],
        [
          'LongIdleTask',
          8,
          {
            type: 'MainThreadLongTask',
            category: 'LongTask',
            startTime: 6,
            endTime: 8,
          },
        ],
      ]
    );

    const drawCalls = flushDrawLog();

    expect(container.firstChild).toMatchSnapshot();
    expect(drawCalls).toMatchSnapshot();

    delete window.devicePixelRatio;
  });

  it('does not render several dot markers in the same position', () => {
    window.devicePixelRatio = 2;

    const { flushDrawLog } = setupWithMarkers(
      { rangeStart: 0, rangeEnd: 15000 },
      [
        // 2 very close dot markers. They shouldn't be drawn both together.
        ['Marker A', 5000, null],
        ['Marker B', 5001, null],
        // This is a longer marker starting at the same place, it should always be drawn
        ['Marker C', 5001, { startTime: 5001, endTime: 7000 }],
      ]
    );

    const drawCalls = flushDrawLog();

    // We filter on height to get only 1 relevant fillRect operation for each marker.
    const fillRectOperations = drawCalls.filter(
      ([operation, , , , height]) => operation === 'fillRect' && height > 1
    );

    // Here 2 markers should be drawn: the first dot, and the long marker.
    expect(fillRectOperations).toHaveLength(2);
    expect(
      fillRectOperations.every(
        ([, , , width]) => width >= MIN_MARKER_WIDTH / window.devicePixelRatio
      )
    ).toBe(true);

    delete window.devicePixelRatio;
  });

  describe('displays context menus', () => {
    beforeEach(() => {
      // Always use fake timers when dealing with context menus.
      jest.useFakeTimers();
      // We will be hovering over element with a tooltip. It requires root overlay
      // element to be present in DOM
      addRootOverlayElement();
    });

    afterEach(removeRootOverlayElement);

    it('when right clicking on a marker', () => {
      const { rightClick, getContextMenu, clickOnMenuItem } = setupWithMarkers(
        { rangeStart: 0, rangeEnd: 10 },
        [['Marker A', 0, { startTime: 0, endTime: 10 }]]
      );

      // The "Marker A" marker is drawn from 0,0 to 5,200.
      rightClick({ x: 50, y: 2 });

      jest.runAllTimers();

      expect(getContextMenu()).toHaveClass('react-contextmenu--visible');

      clickOnMenuItem('Copy');

      expect(copy).toHaveBeenLastCalledWith('Marker A');
      expect(getContextMenu()).not.toHaveClass('react-contextmenu--visible');

      jest.runAllTimers();

      expect(document.querySelector('react-contextmenu')).toBeFalsy();
    });

    it('when right clicking on markers in a sequence', () => {
      const { rightClick, getContextMenu, clickOnMenuItem } = setupWithMarkers(
        { rangeStart: 0, rangeEnd: 10 },
        [
          ['Marker A', 0, { startTime: 0, endTime: 3 }],
          ['Marker B', 0, { startTime: 6, endTime: 10 }],
        ]
      );

      // The "Marker A" marker is drawn from 0,0 to 5,60.
      rightClick({ x: 30, y: 2 });
      jest.runAllTimers();

      // The "Marker B" marker is drawn from 0,120 to 5,200.
      rightClick({ x: 160, y: 2 });
      jest.runAllTimers();

      expect(getContextMenu()).toHaveClass('react-contextmenu--visible');

      clickOnMenuItem('Copy');
      expect(copy).toHaveBeenLastCalledWith('Marker B');

      expect(getContextMenu()).not.toHaveClass('react-contextmenu--visible');

      jest.runAllTimers();
      expect(document.querySelector('react-contextmenu')).toBeFalsy();
    });

    it('and still highlights other markers when hovering them', () => {
      const {
        rightClick,
        mouseOver,
        flushDrawLog,
        getContextMenu,
      } = setupWithMarkers({ rangeStart: 0, rangeEnd: 10 }, [
        ['Marker A', 0, { startTime: 0, endTime: 3 }],
        ['Marker B', 0, { startTime: 6, endTime: 10 }],
      ]);

      // The "Marker A" marker is drawn from 0,0 to 5,60.
      rightClick({ x: 30, y: 2 });
      expect(getContextMenu()).toHaveClass('react-contextmenu--visible');

      flushDrawLog();
      // The "Marker B" marker is drawn from 0,120 to 5,200.
      mouseOver({ x: 160, y: 1 });

      const drawCalls = flushDrawLog();

      // Expect that we have one marker with hovered color
      const callsWithHoveredColor = drawCalls.filter(
        ([, argument]) => argument === overlayFills.HOVERED
      );
      expect(callsWithHoveredColor).toHaveLength(1);
    });
  });
});
