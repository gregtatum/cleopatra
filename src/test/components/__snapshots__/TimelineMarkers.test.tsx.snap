// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineMarkers renders correctly 1`] = `
<div
  class="timelineMarkers selected"
  data-testid="TimelineMarkersOverview"
>
  <div
    class="react-contextmenu-wrapper"
  >
    <canvas
      class="timelineMarkersCanvas"
      height="300"
      width="200"
    />
  </div>
</div>
`;

exports[`TimelineMarkers renders correctly 2`] = `
Array [
  Array [
    "clearRect",
    0,
    0,
    200,
    300,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    1,
    0,
    131.33333333333331,
    1,
  ],
  Array [
    "fillRect",
    0,
    1,
    133.33333333333331,
    4,
  ],
  Array [
    "fillRect",
    1,
    5,
    131.33333333333331,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    1,
    0,
    131.33333333333331,
    1,
  ],
  Array [
    "fillRect",
    0,
    1,
    133.33333333333331,
    4,
  ],
  Array [
    "fillRect",
    1,
    5,
    131.33333333333331,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    28,
    0,
    144.66666666666666,
    1,
  ],
  Array [
    "fillRect",
    27,
    1,
    146.66666666666666,
    4,
  ],
  Array [
    "fillRect",
    28,
    5,
    144.66666666666666,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    28,
    0,
    51.333333333333336,
    1,
  ],
  Array [
    "fillRect",
    27,
    1,
    53.333333333333336,
    4,
  ],
  Array [
    "fillRect",
    28,
    5,
    51.333333333333336,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    68,
    0,
    131.33333333333331,
    1,
  ],
  Array [
    "fillRect",
    67,
    1,
    133.33333333333331,
    4,
  ],
  Array [
    "fillRect",
    68,
    5,
    131.33333333333331,
    1,
  ],
  Array [
    "set fillStyle",
    "black",
  ],
  Array [
    "fillRect",
    81,
    0,
    24.666666666666668,
    1,
  ],
  Array [
    "fillRect",
    80,
    1,
    26.666666666666668,
    4,
  ],
  Array [
    "fillRect",
    81,
    5,
    24.666666666666668,
    1,
  ],
  Array [
    "scale",
    1,
    1,
  ],
]
`;
